lightning/src/ln/msgs.rs

```diff
@@ -983,11 +1006,19 @@ impl Writeable for OnionHopData {
 					(6, short_channel_id)
 				});
 			},
-			OnionHopDataFormat::FinalNode => {
-				encode_varint_length_prefixed_tlv!(w, {
-					(2, HighZeroBytesDroppedVarInt(self.amt_to_forward)),
-					(4, HighZeroBytesDroppedVarInt(self.outgoing_cltv_value))
-				});
+			OnionHopDataFormat::FinalNode { ref payment_data } => {
+				if let &Some(ref final_data) = payment_data {
+					encode_varint_length_prefixed_tlv!(w, {
+						(2, HighZeroBytesDroppedVarInt(self.amt_to_forward)),
+						(4, HighZeroBytesDroppedVarInt(self.outgoing_cltv_value)),
+						(8, final_data)
+					});
+				} else {
+					encode_varint_length_prefixed_tlv!(w, {
+						(2, HighZeroBytesDroppedVarInt(self.amt_to_forward)),
+						(4, HighZeroBytesDroppedVarInt(self.outgoing_cltv_value))
+					});
+				}
```

Comment:
Could you use pattern matching for `payment_data` so as to clearly delineated the cases?

```rust
OnionHopDataFormat::FinalNode { payment_data: Some(ref payment_data) } => {
    // ...
},
OnionHopDataFormat::FinalNode { payment_data: None } => {
    // ...
},
```